File: MorySnip/Classes/ToolStripPalitra.vb
Imports System.ComponentModel

Public Class PalitraEventArgs
    Inherits EventArgs

    Public Property NewColor As Color
    Public Property OldColor As Color

    Public Sub New(NewColor As Color, OldColor As Color)
        Me.NewColor = NewColor
        Me.OldColor = OldColor
    End Sub

    Public Sub New()

    End Sub
End Class

<DefaultEvent("ColorChanged")>
Public Class ToolStripPalitra
    Inherits System.Windows.Forms.ToolStripItem

#Region "Palitra"
    Private Shared _Palitra As Color(,) = Nothing
    Public Shared ReadOnly Property Palitra As Color(,)
        Get
            If _Palitra Is Nothing Then
                _Palitra = New Color(11, 360) {}
                For l As Integer = 0 To 10
                    For i As Integer = 0 To 360 - 1
                        _Palitra(l, i) = RGBHSL.HSL_to_RGB(i / 365, 1 - l / 10, 1)
                    Next
                Next
            End If

            Return _Palitra
        End Get
    End Property
#End Region

    Public Event ColorChanged(sender As Object, e As PalitraEventArgs)
    Public Event Color1Changed(sender As Object, e As PalitraEventArgs)
    Public Event Color2Changed(sender As Object, e As PalitraEventArgs)

    Private _Color1 As Color = Color.Red
    <DefaultValue(GetType(Color), "255, 0, 0")>
    Public Property Color1 As Color
        Get
            Return Me._Color1
        End Get
        Set(value As Color)
            If Not Me._Color1 = value Then
                RaiseEvent ColorChanged(Me, New PalitraEventArgs(value, Me._Color1))
                RaiseEvent Color1Changed(Me, New PalitraEventArgs())
                Me._Color1 = value
                Me.Parent.Refresh()
            End If
        End Set
    End Property

    Private _Color2 As Color = Color.Red
    <DefaultValue(GetType(Color), "255, 0, 0")>
    Public Property Color2 As Color
        Get
            Return Me._Color2
        End Get
        Set(value As Color)
            If Not Me._Color2 = value Then
                RaiseEvent ColorChanged(Me, New PalitraEventArgs(value, Me._Color2))
                RaiseEvent Color2Changed(Me, New PalitraEventArgs())
                Me._Color2 = value
                Me.Parent.Refresh()
            End If
        End Set
    End Property

    Dim Color1Rect As New Rectangle(5, 5, 20, 20)
    Dim Color2Rect As New Rectangle(15, 15, 20, 20)

    Protected Overrides Sub OnPaint(e As PaintEventArgs)
        MyBase.OnPaint(e)

        Dim g As Graphics = e.Graphics
        'g.Clear(Color.Red)

        g.FillRectangle(New SolidBrush(Me.Color2), Me.Color2Rect)
        g.DrawRectangle(Pens.Black, Me.Color2Rect)
        g.FillRectangle(New SolidBrush(Me.Color1), Me.Color1Rect)
        g.DrawRectangle(Pens.Black, Me.Color1Rect)

        For l As Integer = 0 To 10
            For i As Integer = 0 To 359 Step 3
                g.FillRectangle(New SolidBrush(Palitra(l, i)), 70 + i, 5 + 3 * l, 3, 3)
            Next
        Next

        g.DrawRectangle(Pens.Black, New Rectangle())
    End Sub

    Private Function GetColorByPoint(p As Point) As Color
        Dim colors = Palitra
        Dim i = (p.Y - 5) \ 3, j = (p.X - 70)

        If i >= 0 AndAlso i < colors.GetLength(0) AndAlso j >= 0 AndAlso j < colors.GetLength(1) Then
            Return colors(i, j)
        Else
            Return Color.Black
        End If
    End Function

    Private Sub SelectColor1(l As Point)
        Me.Color1 = Me.GetColorByPoint(l)
    End Sub

    Private Sub SelectColor2(l As Point)
        Me.Color2 = Me.GetColorByPoint(l)
    End Sub

    Protected Overrides Sub OnMouseDown(e As MouseEventArgs)
        If e.Button = MouseButtons.Left Then Me.SelectColor1(e.Location)
        If e.Button = MouseButtons.Right Then Me.SelectColor2(e.Location)

        MyBase.OnMouseDown(e)
    End Sub

    Protected Overrides Sub OnMouseMove(mea As MouseEventArgs)
        If mea.Button = MouseButtons.Left Then Me.SelectColor1(mea.Location)
        If mea.Button = MouseButtons.Right Then Me.SelectColor2(mea.Location)

        MyBase.OnMouseMove(mea)
    End Sub

    Public Sub New()
        'Me.Container.Add(New Button())
    End Sub
End Class

File: MorySnip/Classes/ToolStripPalitra.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;

namespace Morysoft.MorySnip;

public class PalitraEventArgs : EventArgs
{
    public Color NewColor { get; set; }

    public Color OldColor { get; set; }

    public PalitraEventArgs(Color NewColor, Color OldColor)
    {
        this.NewColor = NewColor;
        this.OldColor = OldColor;
    }

    public PalitraEventArgs()
    {
    }
}

public delegate void ColorChangedEventHandler(object sender, PalitraEventArgs e);

public delegate void Color1ChangedEventHandler(object sender, PalitraEventArgs e);

public delegate void Color2ChangedEventHandler(object sender, PalitraEventArgs e);

[DefaultEvent("ColorChanged")]
public class ToolStripPalitra : ToolStripItem
{
    private static Color[,]? palitra;

    private static Color[,] Palitra
    {
        get
        {
            if (palitra == null)
            {
                palitra = new Color[12, 361];
                for (int l = 0; l <= 10; l++)
                {
                    for (int i = 0; i <= 360 - 1; i++)
                    {
                        palitra[l, i] = RGBHSL.HSL_to_RGB(i / (double)365, 1 - l / (double)10, 1);
                    }
                }
            }

            return palitra;
        }
    }

    public event ColorChangedEventHandler? ColorChanged;

    public event Color1ChangedEventHandler? Color1Changed;

    public event Color2ChangedEventHandler? Color2Changed;

    private Color color1 = Color.Red;
    private Color color2 = Color.Red;

    private Rectangle Color1Rect = new(5, 5, 20, 20);
    private Rectangle Color2Rect = new(15, 15, 20, 20);

    [DefaultValue(typeof(Color), "255, 0, 0")]
    public Color Color1
    {
        get => this.color1;
        set
        {
            if (this.color1 == value)
            {
                return;
            }

            if (this.color2 == this.color1)
            {
                this.color2 = value;
                Color2Changed?.Invoke(this, new PalitraEventArgs());
            }

            var oldColor = this.color1;

            this.color1 = value;
            Color1Changed?.Invoke(this, new PalitraEventArgs());

            ColorChanged?.Invoke(this, new PalitraEventArgs(value, oldColor));

            this.Parent.Refresh();
        }
    }

    [DefaultValue(typeof(Color), "255, 0, 0")]
    public Color Color2
    {
        get => this.color2;
        set
        {
            if (this.color2 == value)
            {
                return;
            }

            var oldColor = this.color2;

            this.color2 = value;
            ColorChanged?.Invoke(this, new PalitraEventArgs(value, oldColor));
            Color2Changed?.Invoke(this, new PalitraEventArgs());
            this.Parent.Refresh();
        }
    }

    protected override void OnPaint(PaintEventArgs e)
    {
        base.OnPaint(e);

        var g = e.Graphics;
        // g.Clear(Color.Red)

        g.FillRectangle(new SolidBrush(this.Color2), this.Color2Rect);
        g.DrawRectangle(Pens.Black, this.Color2Rect);
        g.FillRectangle(new SolidBrush(this.Color1), this.Color1Rect);
        g.DrawRectangle(Pens.Black, this.Color1Rect);

        for (int l = 0; l <= 10; l++)
        {
            for (int i = 0; i <= 359; i += 3)
            {
                g.FillRectangle(new SolidBrush(Palitra[l, i]), 70 + i, 5 + 3 * l, 3, 3);
            }
        }

        g.DrawRectangle(Pens.Black, new Rectangle());
    }

    private static Color GetColorByPoint(Point p)
    {
        var colors = Palitra;
        int i = (p.Y - 5) / 3;
        int j = p.X - 70;

        return i >= 0 && i < colors.GetLength(0) && j >= 0 && j < colors.GetLength(1)
            ? colors[i, j]
            : Color.Black;
    }

    private void SelectColor1(Point point) => this.Color1 = GetColorByPoint(point);

    private void SelectColor2(Point point) => this.Color2 = GetColorByPoint(point);

    protected override void OnMouseDown(MouseEventArgs e)
    {
        if (e.Button == MouseButtons.Left)
        {
            this.SelectColor1(e.Location);
        }

        if (e.Button == MouseButtons.Right)
        {
            this.SelectColor2(e.Location);
        }

        base.OnMouseDown(e);
    }

    protected override void OnMouseMove(MouseEventArgs e)
    {
        if (e.Button == MouseButtons.Left)
        {
            this.SelectColor1(e.Location);
        }

        if (e.Button == MouseButtons.Right)
        {
            this.SelectColor2(e.Location);
        }

        base.OnMouseMove(e);
    }

    public ToolStripPalitra()
    {
    }
}
